segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -123,12 +130,32 @@
      * {@link ContainerEventProcessor.EventProcessor} based on its name.
      */
     @VisibleForTesting
-    static Function<String, CompletableFuture<DirectSegmentAccess>> getOrCreateInternalSegment(SegmentContainer container,
-                                                                                               MetadataStore metadataStore,
-                                                                                               Duration timeout) {
-        return s -> metadataStore.registerPinnedSegment(getEventProcessorSegmentName(container.getId(), s),
-                SYSTEM_CRITICAL_SEGMENT, null, timeout) // Segment should be pinned.
-                .thenCompose(l -> container.forSegment(getEventProcessorSegmentName(container.getId(), s), timeout));
+    Function<String, CompletableFuture<DirectSegmentAccess>> getOrCreateInternalSegment(SegmentContainer container,
+                                                                                        MetadataStore metadataStore,
+                                                                                        Duration timeout) {
+        return s -> tryCreateAndPinInternalSegment(container, metadataStore, s, timeout)
+                .thenCompose(v -> container.forSegment(getEventProcessorSegmentName(container.getId(), s), timeout));
+    }
+
+    private CompletableFuture<Void> tryCreateAndPinInternalSegment(SegmentContainer container, MetadataStore metadataStore,
+                                                                   String eventProcessorName, Duration timeout) {
+        CompletableFuture<EventProcessor> eventProcessorFuture = this.eventProcessorMap.get(eventProcessorName);
+        if (eventProcessorFuture == null || !eventProcessorFuture.isDone()) {
+            String segmentName = getEventProcessorSegmentName(container.getId(), eventProcessorName);
+            return container.createStreamSegment(segmentName, SYSTEM_CRITICAL_SEGMENT, null, timeout)
+                    .thenAccept(v -> log.info("{}: Created internal Segment {}.", this.traceObjectId, segmentName))
+                    .exceptionally(e -> {
+                        val ex = Exceptions.unwrap(e);
+                        if (e.getCause() instanceof StreamSegmentExistsException) {
+                            log.info("{}: ContainerEventProcessor Segment {} already exists.", this.traceObjectId, segmentName);
+                            return null;
+                        }
+                        log.error("{}: Unexpected problem creating internal Segment {}.", this.traceObjectId, segmentName, ex);
+                        throw new CompletionException(ex);
+                    })
+                    .thenAccept(v -> metadataStore.pinSegmentToMemory(getEventProcessorSegmentName(container.getId(), eventProcessorName), timeout)); // Pin the Segment in memory.
```

Comment:
Need to move 
`.thenAccept(v -> metadataStore.pinSegmentToMemory(getEventProcessorSegmentName(container.getId(), eventProcessorName), timeout));`

before `.exceptionally`

Otherwise the metadataStore.pinSegmentToMemory is called even after StreamSegmentExistsException

Reply:
goal here is to avoid calling metadataStore.pinSegmentToMemory on every call

Reply:
Also note that - in general, we need to be careful with putting code that returns CompletableFuture inside synchronize blocks.
Because it is only the creation of CompletableFuture that runs in synchronize not the actual execution of lambda that eventually completes the future. In other words it is possible that code that we really want to run in synchronized block ends up running unprotected outside the block.

Reply:
>Otherwise the metadataStore.pinSegmentToMemory is called even after StreamSegmentExistsException

This is exactly what we want, because the fact that the Segment exists does not mean that it is pinned to memory. Irrespective of whether the Segment is new or not, we want to pin it.

Reply:
> goal here is to avoid calling metadataStore.pinSegmentToMemory on every call.

`pinSegmentToMemory` is not called every time. If you look at the code, there is an `if` clause that prevents any further call to this method (and to the `createStreamSegment` method) once the event processor has been initialized.

Reply:
> In other words it is possible that code that we really want to run in synchronized block ends up running unprotected outside the block.

The only thing I would like to synchronize is the access to the map, that's it. If the map does not contain the expected future or it is not complete, we will trigger the segment creating and pin it to memory. If it exist and it is complete, we are doing nothing. In the worst case, what could happen is to issue a Segment creation and pin just after the Event Processor is removed from the map (e.g., during shutdown?), which I believe is harmless: either the operation will succeed (i.e., finding that the Segment already exists and not pinning it to memory, because it is already pinned) or not (failing the operation because the container is shut down already), but that's it. I don't see any other downside of this approach at the moment.